Create drake/common subdirectory and put NiceTypeName<T> in it.

sherm1 opened the pull request:
Per discussion in #2153 (need for `common` subdirectory) and Reviewable discussion in #2112 (need for demangled type names in error messages):

- Adds `drake/common` subdirectory to be used instead of the overloaded `drake/util` directory. A README describes how the subdirectory should be used.
- Initially populates `drake/common` with a single class, `NiceTypeName<T>` that provides a static method returning a human-readable, canonicalized string that should be the same on all platforms.

@david-german-tri, I don't think there is any hope for the existing `drake/util` subdirectory, which is really devoted to Matlab. There may be a few C++ classes there we can move to `common`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2172)
<!-- Reviewable:end -->

liangfok commented:
Looks great overall. :thumbsup:  I Just left some minor comments.

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[drake/common/CMakeLists.txt, line 5 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBCewIrGW1IsuZKxjN:-KGBCewIrGW1IsuZKxjO:1005749151) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/CMakeLists.txt#L5)):*
I'm not sure if the rule applies here since this is not technically a function, and it's an extreme nit regardless, but the Google Style guide recommends using a descriptive tense ("Lists all...") rather than an imperative tense ("List all...").

https://google.github.io/styleguide/cppguide.html#Function_Comments

---

*[drake/common/nice_type_name.h, line 13 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBIkR5Yx8E1ynhJr7y:-KGBIkR6AWq4vZvYJxMR:1818678580) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L13)):*
"demangle" --> "demangles"

---

*[drake/common/nice_type_name.h, line 17 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBJRq1tIAriQ_4CGYO:-KGBJRq1tIAriQ_4CGYP:-1965243124) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L17)):*
I do like your compact Doxygen block style. If no one objects, let's propose to make it standard throughout Drake in https://github.com/RobotLocomotion/drake/issues/2051.

---

*[drake/common/nice_type_name.h, line 18 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBJBDQJ2lTOpVIEzT6:-KGBJBDQJ2lTOpVIEzT7:-778287740) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L18)):*
Is the `@cond` and `@endcond` a Doxygen thing? What purposes does it serve? Should the rest of Drake be updated to use it as well when using similar `*_EXPORT` macros?

---

*[drake/common/nice_type_name.h, line 64 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBKrPS4e4G1VzLY1Rb:-KGBKrPS4e4G1VzLY1Rc:313371865) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L64)):*
I'm having trouble understanding how it's only done once per instantiation of `NiceTypeName<T>`. Is there a cache being maintained somewhere that I'm missing? Is it because of the use of the `static const` keyword? If so, very clever use of that keyword!

---

*[drake/common/nice_type_name.h, line 73 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBNQ1u6UaBKpKB6w2B:-KGBNQ1u6UaBKpKB6w2C:-1401988233) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L73)):*
I see that getting the human-readable name of a type incurs some memory overhead. Is it worth structuring this feature so the memory overhead is only incurred on debug builds and not on release builds?

---

*[drake/common/README.md, line 17 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBHFqVOj7-4ZLeV-DD:-KGBHFqWvTkPSkYTJfCR:-611432709) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/README.md#L17)):*
Interesting rule to prefix the name of all unit test executables with "test-". Should this be made a general rule throughout Drake? If so, let's add it to https://github.com/RobotLocomotion/drake/issues/1973.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-:-KGBLrk6usysS-MhLQkD:-1380042366)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[drake/common/CMakeLists.txt, line 5 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBCewIrGW1IsuZKxjN:-KGD51tTdn2H87Q_Uzc5:1735698004) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/CMakeLists.txt#L5)):*
I was thinking of this as an instruction to future developers rather than as a description of the variable. Let me know if you still want me to change it. 

---

*[drake/common/nice_type_name.h, line 13 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBIkR5Yx8E1ynhJr7y:-KGD5pyiwxtPnZtDsoe3:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L13)):*
Done.

---

*[drake/common/nice_type_name.h, line 17 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBJRq1tIAriQ_4CGYO:-KGD5wZOM1QggmiJ7nAl:-362181657) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L17)):*
I'd like to see it as an allowable format but not an absolute requirement since I'm sure some people are very attached to the old format, and these are just comments anyway.

---

*[drake/common/nice_type_name.h, line 18 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBJBDQJ2lTOpVIEzT6:-KGD6GYh1NURik2Zecqr:1046077902) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L18)):*
Yes. That's conditional inclusion in Doxygen with the condition being "false" so that Doxygen won't see the ugly and meaningless export symbol. Normally this would be done in the Doxyfile by providing an empty expansion for these symbols. However, since the export symbols are generated in Drake I'm not sure how to do that. Rather than sprinkle these around Drake, I filed issue #2177. 

---

*[drake/common/nice_type_name.h, line 64 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBKrPS4e4G1VzLY1Rb:-KGDAFEio1H8yu9C9kxu:-8381631) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L64)):*
Right. C++11 guarantees thread safe single execution of static initializers in functions. The generated code includes the equivalent if `if (!initialized) {...}`. Each instantiation of `NiceTypeName<xxx_t>` for a particular type shares the same static methods, so the initialization gets done just once (although at runtime the hidden flag must be checked with every call).

---

*[drake/common/nice_type_name.h, line 73 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBNQ1u6UaBKpKB6w2B:-KGDB9sYK1ouEm5jNY7_:-687994283) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L73)):*
The overhead only exists if `NiceTypeName<T>` is instantiated for some T. That could happen for good reason in a Release build, but one could also avoid instantiation except in Debug builds -- that doesn't have to be dealt with here.

---

*[drake/common/README.md, line 17 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBHFqVOj7-4ZLeV-DD:-KGDBbgms_TgEc1743iw:2089129883) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/README.md#L17)):*
It is certainly convenient for the Visual Studio IDE, and most of the existing test cases begin with the word "test" (without the hyphen). Can anyone comment on how these look in CLion?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@liangfok, thanks for the review. Updated.

---

Review status: 4 of 7 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-:-KGDCzT_jTzrJResBbCa:-80753914)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 2 of 3 files at r2.
Review status: 6 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/common/CMakeLists.txt, line 5 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBCewIrGW1IsuZKxjN:-KGDEocRMRFaDkyDoCSF:-725923228) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/CMakeLists.txt#L5)):*
No need to change. Just wanted to run the idea by you.

---

*[drake/common/nice_type_name.h, line 17 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBJRq1tIAriQ_4CGYO:-KGDE0ZoplCFhbv_WElV:1127309162) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L17)):*
OK for making it one of many allowable formats. I was just thinking standardizing on a single format throughout Drake would offer greater conceptual clarity in terms of making the code appear as if it were written by a single person. But I guess it doesn't matter since you're right, these are just comments and we have way larger fires to put out at the moment.

---

*[drake/common/nice_type_name.h, line 64 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBKrPS4e4G1VzLY1Rb:-KGDEO_riySMsxbZijsx:-1671032991) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L64)):*
TIL. Thanks.

---

*[drake/common/nice_type_name.h, line 73 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBNQ1u6UaBKpKB6w2B:-KGDETwa-WZxFyA_fjjI:-782328506) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L73)):*
OK, no changes necessary.

---

*[drake/common/README.md, line 17 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBHFqVOj7-4ZLeV-DD:-KGDEZCt3M-ZR6tfVKPu:-610980602) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/README.md#L17)):*
I haven't used CLion yet but having all unit tests start with "test-" would organize things better in Sublime Text as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-:-KGDEmSBbx1jPg27c5-M:-1868698381)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 5 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/common/nice_type_name.cc, line 27 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEJkqOkOnYd9mp8rRy:-KGEJkqOkOnYd9mp8rRz:-2133893770) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L27)):*
If you use `namespace drake {` and `namespace common {` around the whole file, there's less risk some future edit will dump stuff in global namespace.

---

*[drake/common/nice_type_name.cc, line 66 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEJx4FGGB7VuAeKKDP:-KGEJx4FGGB7VuAeKKDQ:-551481565) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L66)):*
std::string, here and elsewhere (please run cpplint!)

---

*[drake/common/nice_type_name.h, line 17 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBJRq1tIAriQ_4CGYO:-KGEK2T0ZpgFoiUXH1xn:1414567959) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/nice_type_name.h#L17)):*
I agree with @liangfok in all respects: mandating a style is low priority, but this style is nice!

---

*[drake/common/README.md, line 17 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBHFqVOj7-4ZLeV-DD:-KGEKU4tHGChmWFr25RL:-182166344) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/README.md#L17)):*
The style guide calls for `_test.cc`, and that's what we've been doing elsewhere: https://google.github.io/styleguide/cppguide.html#File_Names

It also implicitly calls for keeping the tests in the same directory as the source: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

These two practices combined make it easy to find the tests for the code you're looking at right now.

---

*[drake/common/test/test-nice_type_name.cc, line 32 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGELa1L2urp49oPhIB0:-KGELa1L2urp49oPhIB1:519447435) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/test/test-nice_type_name.cc#L32)):*
is GTEST_TEST different from simply TEST?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 5 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/common/nice_type_name.cc, line 27 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEJkqOkOnYd9mp8rRy:-KGESRaOdZD7kY2VaKGy:-39621287) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L27)):*
Hmmm ... that seems wrong because I don't want this file to add symbols to those namespaces either -- it is just supposed to be defining things that have already been declared. And I don't want to elide the namespaces here -- they should be explicit so it is clear what's being defined. Maybe the whole file should be wrapped in an anonymous namespace? Should that be standard practice for every .cc file?

---

*[drake/common/nice_type_name.cc, line 66 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEJx4FGGB7VuAeKKDP:-KGETlfP2fLwEGwQVDE4:285358612) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L66)):*
I didn't see any complaint from cpplint. I intentionally wrote `using std::string` at the top of this cc file so I could do this. The style guide seemed OK with that as long as it didn't expose the whole namespace, but I'd be happy to change it if you like your strings more explicit.

---

*[drake/common/README.md, line 17 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBHFqVOj7-4ZLeV-DD:-KGEWYUjNDGbOgrjDGet:-184737302) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/README.md#L17)):*
I am happy to change the name and get rid of the `test` subdirectory if you think that is better. However, the style guide seems flexible on these points, e.g. "dir/foo.cc and dir2/foo2.h are usually in the same directory (e.g. base/basictypes_test.cc and base/basictypes.h), but may sometimes be in different directories too." So I think we are free to do whatever makes the most sense here -- happy to defer to you on that. Are you sure this is the best approach? Downsides I see: mixing in tests with source seems "impure" since they are not part of the same target begin built (for example, it makes sense not to build the tests at all sometimes); adds roughly 50% to the number of files in the source directory; test executables don't sort nicely in the IDE; `test` subdirectory is used fairly consistently elsewhere in Drake at the moment and is very common in general; the  `_test.cc` suffix is mildly ambiguous since we also separate class name words with `_` in the files (would be a problem for a class named SmogTest: smog_test.cc could be test case for class Smog). Not sure what the upsides are.

---

*[drake/common/test/test-nice_type_name.cc, line 32 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGELa1L2urp49oPhIB0:-KGEYyH9jQ71MEZOgLNU:436910983) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/test/test-nice_type_name.cc#L32)):*
They are the same thing. `TEST` is a mistake Google made (in violation of its own standards I think) in creating a macro with such a common name that it conflicts with other code. Hence the need for [this kludge](https://github.com/google/googletest/blob/master/googletest/docs/FAQ.md#google-test-defines-a-macro-that-clashes-with-one-defined-by-another-library-how-do-i-deal-with-that). Other mistakes were `SUCCEED` and `FAIL`. I think we should be defining the provided `GTEST_DONT_DEFINE_TEST`etc. macros to avoid the potential for clash with third party or user code. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 5 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/common/nice_type_name.cc, line 27 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEJkqOkOnYd9mp8rRy:-KGEdSqQ-F_FmaGnHWAx:550204348) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L27)):*
I'm pretty sure that if you wrap the whole file in an anonymous namespace, drake::common::DemangleTypeName will become an undefined symbol.  My opinion is that the downsides to wrapping the whole file in `namespace foo {` are pretty small, and it's the approach implied by the examples here: https://google.github.io/styleguide/cppguide.html#Namespaces

---

*[drake/common/nice_type_name.cc, line 66 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEJx4FGGB7VuAeKKDP:-KGEccb-CnuZRjgZj7Be:-824909498) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L66)):*
Oh, I didn't notice the using-decl.  I retract the comment!

---

*[drake/common/README.md, line 17 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBHFqVOj7-4ZLeV-DD:-KGEehBb_jgJbY6VHB_R:-363127847) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/README.md#L17)):*
Regarding the directory - fair points.  I think the main upside is that test files sort next to the source files they test. I more often say "hey, I should look at the tests for this file" than "hey, I should look at a list of all the tests for code in this directory".  I suppose we should have another issue-poll: #2182

Regarding the filename - I think the style guide is fairly explicit about `_test`.  It goes so far as to note deprecated alternatives.

---

*[drake/common/test/test-nice_type_name.cc, line 32 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGELa1L2urp49oPhIB0:-KGEdAs_uxNvTrlgYb0t:-1965538027) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/test/test-nice_type_name.cc#L32)):*
Interesting, didn't know that, and I agree.  Opened #2181.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[drake/common/nice_type_name.cc, line 6 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEs2kkcAb5Nkydbqj5:-KGEs2kkcAb5Nkydbqj6:-1531422366) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L6)):*
I think it is worth asking the question whether a license statement that differs from the top-level LICENSE.TXT should be allowed in-tree.

Examples of files in-tree with differing copyright are few (we have thirdParty, and eiquadprog, and nothing else that I can see).

Or at least, perhaps LICENSE.TXT should be amended with an "unless otherwise noted in the source code" or something.

---

*[drake/common/nice_type_name.cc, line 47 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEwcICOBO_leqkhH9t:-KGEwcIDsOm8db-HiSBD:2140147834) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L47)):*
"Variables of class type with static storage duration are forbidden: they cause hard-to-find bugs due to indeterminate order of construction and destruction."
-- https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

---

*[drake/common/nice_type_name.h, line 19 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEtYSSCPqmTY_kvfSn:-KGEtYSSCPqmTY_kvfSo:100887680) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.h#L19)):*
This may be a losing battle, but feel like namespace-functions should be careful about cropping up in surprising header files.  I think the `path/to/foo_bar.h` file should typically only claim the `path::to::FooBar` name, and not the `path::to::HelperFunction`.  The exception being perhaps things like `path/to/geometry_functions.h` where it's clear it's a bag of namespace-functions, not a class.  Not mixing classes and functions in the same header file is my typical rule of thumb.  This is all in terms of being more obvious to a reader of the calling-code where the function being called is coming from.

In this case, I would prefer to see this function called `drake::common::NiceTypeNameDemangle` or `drake::common::NiceTypeName_detail::DemangleTypeName`, or even better...  make NiceTypeName non-templated, relocate this to be a `NiceTypeName::Demangle(const char*)`static member function, and have remplate Get<T>() methods, not classes, for turning `T` into a typeid.

---

*[drake/common/nice_type_name.h, line 62 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGExZUm8Jkmdt27k7Ly:-KGExZUnLpfnLeX7c6th:1602351887) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.h#L62)):*
Inconsistent verb; s/Attempt/Attempts/.  (All other API comments in this file are third-person present tense.)

---

*[drake/common/nice_type_name.h, line 67 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGExJuU82AP2QSWacLw:-KGExJuU82AP2QSWacLx:1285460672) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.h#L67)):*
Another non-POD static storage duration.

---

*[drake/common/nice_type_name.h, line 77 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGExNXB5N0xl-UXYdh9:-KGExNXB5N0xl-UXYdhA:1285460672) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.h#L77)):*
Another non-POD static storage duration.

---

*[drake/common/README.md, line 19 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGExrwagrYexLcDEmvg:-KGExrwagrYexLcDEmvh:-809855472) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/README.md#L19)):*
Missing newline at end of file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 2 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/common/nice_type_name.cc, line 6 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEs2kkcAb5Nkydbqj5:-KGFhDNxkQrsY9TDPpB6:180602829) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L6)):*
This is commonly done and I think it is OK as long as the license is compatible. Otherwise we can't conveniently grab open source code to hack for our purposes. That would be very disappointing to me since that is the whole purpose of extremely liberal licensing like BSD or Apache over viral licenses (and why I have used those licenses myself). The only thing we can't do is remove the copyright block (although we can relocate it within the file). We are still waiting for a resolution on issue #1805 to establish proper licensing & copyright for Drake. After that the Drake copyright will be at the top of every file pushing these borrowed things down.

I updated LICENSE.TXT as you suggested. Now it reads: `All components of Drake are licensed under the BSD 3-Clause License shown below. Where noted in the source code, some portions may be subject to other permissive, non-viral licenses.` We can revisit when #1805 is resolved.

---

*[drake/common/nice_type_name.cc, line 27 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEJkqOkOnYd9mp8rRy:-KGFgriUw_DAds4SGSBI:935257189) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L27)):*
You're right that wrapping in an anonymous namespace doesn't work. I wrapped the code in `namespace drake { namespace common {` as you suggested. 

---

*[drake/common/nice_type_name.cc, line 47 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEwcICOBO_leqkhH9t:-KGFeZqZOrAcI-VQMtq9:-640151877) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L47)):*
That is referring to namespace-scope static variables because initialization order for them is undefined. The initialization order for function-scope statics is well defined as it says there. The text isn't clear because it jumbles in POD in the same sentence (because they are talking about using a function for initialization) but in fact it doesn't matter what kind of function-scope static variable you have - they all get initialized nicely, in a thread-safe manner, the first time the function is called.

---

*[drake/common/nice_type_name.h, line 19 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEtYSSCPqmTY_kvfSn:-KGFz6BN8BU0W8OjAC8X:1735456161) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.h#L19)):*
Battle won! I made NiceTypeName non-templated, moved Demangle() and Canonicalize() into it as static members, and templatized Get<T>() and GetRaw<T>(). I attempted to fix up all the doxygen and comments to match; please let me know if you see anything I missed.

---

*[drake/common/nice_type_name.h, line 62 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGExZUm8Jkmdt27k7Ly:-KGFykBR7rWqncmcWu8v:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.h#L62)):*
Done.

---

*[drake/common/nice_type_name.h, line 67 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGExJuU82AP2QSWacLw:-KGFmHqtI_UXNW9dsYyP:1216230207) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.h#L67)):*
And still OK!

---

*[drake/common/nice_type_name.h, line 77 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGExNXB5N0xl-UXYdh9:-KGFmLTmnY2h3cmMLKmi:-2086277040) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.h#L77)):*
:+1: Good for me!

---

*[drake/common/README.md, line 17 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGBHFqVOj7-4ZLeV-DD:-KGFnArml2K8Bv4SohSP:1250270886) ([raw file](https://github.com/robotlocomotion/drake/blob/ed075161524c6fc9056ef90487f4bf541da49690/drake/common/README.md#L17)):*
Renamed to nice_type_name_test.cc. Will leave the `test` directory there pending outcome of #2182.

---

*[drake/common/README.md, line 19 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGExrwagrYexLcDEmvg:-KGFn2ZW4FOU-RwWyq4W:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/README.md#L19)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 8 files at r1, 3 of 7 files at r4.
Review status: 5 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/common/nice_type_name.cc, line 6 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEs2kkcAb5Nkydbqj5:-KGG0Rb--QMYDRnvqHTi:-303870841) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L6)):*
Oh, I definitely don't want to prevent us from using Apache-2.0 code within Drake.  I just want us to be careful about copyrights.  Untangling it later is always a huge pain.  I think the LICENSE.txt change here is a step in the right direction.

---

*[drake/common/nice_type_name.cc, line 47 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEwcICOBO_leqkhH9t:-KGG0twjGCskztp7Aby1:-623028993) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L47)):*
I read the google cppguide and cppreference.com carefully before commenting above.  Static local variables still have static storage duration, and are of class type (as opposed to POD type), and so are verboten per the guide.  _("Objects with static storage duration, including ... function static variables, must be Plain Old Data (POD) ...")._

I agree that (as of C++11) the initialization order here is repeatably well-defined (assuming the std::string implementation uses only Static (Constant or Zero) Initialization, and not  Dynamic Initialization)  However, I believe that destructor-order is still indeterminate, which is the reason for the rule as written.  And you might imagine that something like NiceTypeName is the kind of that that might be invoked during program exit.

---

*[drake/common/nice_type_name.h, line 36 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGG4gtzOkMdqb7w5rBK:-KGG4gtzOkMdqb7w5rBL:1158967617) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L36)):*
:thumbsup: I like that Get() is atop the file now -- the thing that people should use.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This comment requires no change, but FTR, I think `NiceTypeName::Get<..>()` is so useful it should be pushed upstream to the C++ standard library (related: https://github.com/RobotLocomotion/drake/issues/2153#issuecomment-213209725).

---

Reviewed 7 of 7 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-:-KGHLHSg0Aq5yYEYzxdA:1529760097)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/common/nice_type_name.cc, line 27 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEJkqOkOnYd9mp8rRy:-KGHzPOUPQSAGXwqS6ud:-1819650420) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L27)):*
That makes the namespacing here redundant, but if you like it I have no objections.

---

*[drake/common/nice_type_name.cc, line 47 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEwcICOBO_leqkhH9t:-KGHzamd5qJJipSFwksj:-1368921124) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L47)):*
Good catch @jwnimmer-tri.  I think this is one of the few, surprising cases where an unmanaged pointer is appropriate:  `static const std::array<SPair, 8>* subs = new std::array<SPair, 8>{...}`

(I'm about 95% sure this doesn't result in unreachable memory, but if you go this route please valgrind the tests to confirm!)

---

*[drake/common/nice_type_name.h, line 51 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGI2I-kP-dMtKoIkzqd:-KGI2I-kP-dMtKoIkzqe:1167570579) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L51)):*
`private:` ? 

---

*[drake/common/nice_type_name.h, line 58 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGI1lFtugobzihbdbR6:-KGI1lFus7ckdTzchOAi:503452439) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L58)):*
nit: whitespace

---

*[drake/common/nice_type_name.h, line 66 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGI39ZwcZy3MjAPRnRp:-KGI39ZwcZy3MjAPRnRq:-943440259) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L66)):*
rvalue references are technically verboten, and as far as I can tell this isn't ever called except from the lvalue version below.  Can this just be deleted?

https://google.github.io/styleguide/cppguide.html#Rvalue_references

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/common/nice_type_name.cc, line 47 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEwcICOBO_leqkhH9t:-KGIKFYaOWnAfRM193cc:987604600) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L47)):*
Yup, I agree that is an acceptable fix.  (Personally, I would have made it a non-static local.  Premature optimization and all that.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/common/nice_type_name.cc, line 47 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEwcICOBO_leqkhH9t:-KGINPRxN5H_S6_paGKo:1779298308) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L47)):*
+1, that works too and looks less scary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/common/nice_type_name.h, line 56 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGKkA_9gCbEGJlnQgy5:-KGKkA_9gCbEGJlnQgy6:1490896586) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L56)):*
FWIW, apparently we could teach Doxygen to automatically ignore the DRAKECOMMON_EXPORT tokens via its PREDEFINED configuration option.
https://www.stack.nl/~dimitri/doxygen/manual/preprocessing.html

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/common/nice_type_name.h, line 56 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGKkA_9gCbEGJlnQgy5:-KGKodzN6uNIwgJdjdpj:718469516) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L56)):*
Thanks, @jwnimmer-tri. I added that link to issue #2177. It still isn't clear to me how to do this with generated export symbols although we could just list the ones we know about in the Doxyfile.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/common/nice_type_name.h, line 29 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGMqPYKe6EK7aknprOK:-KGMqPYL_KmB_VYdBh4t:675228392) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L29)):*
The default constructor is still enabled, but should not be -- since nobody should instantiate this class.

(Sorry for the slow trickle of comments.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/common/nice_type_name.h, line 29 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGMqPYKe6EK7aknprOK:-KGNhiUWUfrnsbH3Agzf:-1036489060) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L29)):*
Good point, @jwnimmer-tri. I'll delete the constructor and other junk. But do you think it is odd to use a class here at all? I share your distaste for free functions so I like collecting these into a class but since it is all static there is no object here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/common/nice_type_name.h, line 29 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGMqPYKe6EK7aknprOK:-KGNjFEw-SQzdL-Q6iSF:186584435) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L29)):*
I think the design is great!  Putting static utility methods into a class is great!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 5 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/common/nice_type_name.cc, line 6 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEs2kkcAb5Nkydbqj5:-KGP8tmzbRLr7j-Y4Hoa:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L6)):*
Done.

---

*[drake/common/nice_type_name.cc, line 47 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGEwcICOBO_leqkhH9t:-KGPDO9K_EApiZtm8J0g:1329333675) ([raw file](https://github.com/robotlocomotion/drake/blob/a20b87e298e529040a3ea916d9ad4038aac19364/drake/common/nice_type_name.cc#L47)):*
I made it a non-static local as you suggested. The Canonicalize() method gets called only once per type and the regex table construction is immediately followed by expensive string processing so I doubt this will ever cause a problem. Definitely premature optimization -- thanks.

---

*[drake/common/nice_type_name.h, line 29 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGMqPYKe6EK7aknprOK:-KGP9A1G5lizKwIiIZ4N:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L29)):*
Done.

---

*[drake/common/nice_type_name.h, line 51 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGI2I-kP-dMtKoIkzqd:-KGP9Csr92CuXaMXBm_X:210008299) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L51)):*
I decided that GetRaw() was ill-motivated and nuked it altogether!

---

*[drake/common/nice_type_name.h, line 58 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGI1lFtugobzihbdbR6:-KGPEFLLaiAy7aY8oW3d:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L58)):*
Done.

---

*[drake/common/nice_type_name.h, line 66 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGI39ZwcZy3MjAPRnRp:-KGPEHqVfNWN9NtqQfNr:-822944050) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L66)):*
I agree this isn't needed here so I deleted it. \<rant>In general I think rvalue references are awesome, one of the most useful C++11 improvements. E.g. the effect on the performance of `std::vector` resizing was **YUGE**. I disagree with the style guide on this point. Its argument for this is particularly weak: "Rvalue references are a relatively new feature (introduced as part of C++11), and not yet widely understood. Rules like reference collapsing, and automatic synthesis of move constructors, are complicated." Now standardized for 5 years, I don't think "relatively new" applies and I certainly don't think the rules are unreasonably complicated (by C++ standards!).\</rant>

---

*[drake/common/nice_type_name.h, line 50 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGPFxAoaCbbDli5orRT:-KGPFxAoaCbbDli5orRU:566207494) ([raw file](https://github.com/robotlocomotion/drake/blob/0e57ea72e81b58d3a44c965569cadcf5c859e8d0/drake/common/nice_type_name.h#L50)):*
Please review this change. I want these strings to persist and I agree with @jwnimmer-tri that they may be useful even during program tear-down.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@jwnimmer-tri, @david-german-tri I think I responded to everything -- please re-review.

---

Review status: 5 of 8 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-:-KGPGqcpbKCD22E_Oa8o:-204298105)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 8 files at r1, 2 of 3 files at r6.
Review status: 7 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/common/nice_type_name.h, line 50 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGPFxAoaCbbDli5orRT:-KGPRDe4NsxiOcldLx6S:-510466975) ([raw file](https://github.com/robotlocomotion/drake/blob/0e57ea72e81b58d3a44c965569cadcf5c859e8d0/drake/common/nice_type_name.h#L50)):*
:+1: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-:-KGPRkJKZJVUQ221KtCw:1417088721)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 2 of 8 files at r1, 4 of 7 files at r4, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/common/nice_type_name.h, line 51 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGI2I-kP-dMtKoIkzqd:-KGS98qd9H3P706NwGTH:-390244029) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L51)):*
Can `Demangle` and `Canonicalize` be private also?

---

*[drake/common/README.md, line 17 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGS9leljtphEfvIDNjt:-KGS9lemlDMRcoB-5VLE:-1737664) ([raw file](https://github.com/robotlocomotion/drake/blob/0e57ea72e81b58d3a44c965569cadcf5c859e8d0/drake/common/README.md#L17)):*
out-of-date

---

*[drake/common/test/nice_type_name_test.cc, line 8 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGS9xNeAD1ZGqfL239l:-KGS9xNeAD1ZGqfL239m:1167644049) ([raw file](https://github.com/robotlocomotion/drake/blob/0e57ea72e81b58d3a44c965569cadcf5c859e8d0/drake/common/test/nice_type_name_test.cc#L8)):*
nit: out-of-order

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
A question for experienced Google-stylers (@david-german-tri et al): I named the primary method `Get()`. Should this be `get()`? As I understand the choice it depends on performance. In this case the method is expensive the first time it is called for a given type, but then just returns a reference to an existing string after that.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-:-KGTVXOajhbYXq0fdT3W:-764728619)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
When in doubt, capitalize.  I think `Get` is fine.

---

Review status: 6 of 8 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-:-KGU9E7jFAowJrQxvV9o:1264404805)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 6 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/nice_type_name.h, line 51 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGI2I-kP-dMtKoIkzqd:-KGU5gNa03rn9mpC6Dli:-241450055) ([raw file](https://github.com/robotlocomotion/drake/blob/76a3ccdb08436ad86ce236a42e47dde7008b84a7/drake/common/nice_type_name.h#L51)):*
I would rather keep them public since they do useful, non-trivial, and easily describable tasks related to making nice type names.

---

*[drake/common/README.md, line 17 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGS9leljtphEfvIDNjt:-KGU8ETjPIFiAeZlKn4_:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/0e57ea72e81b58d3a44c965569cadcf5c859e8d0/drake/common/README.md#L17)):*
Done.

---

*[drake/common/test/nice_type_name_test.cc, line 8 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-KGS9xNeAD1ZGqfL239l:-KGU9MrMQFzjaUBORJJd:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/0e57ea72e81b58d3a44c965569cadcf5c859e8d0/drake/common/test/nice_type_name_test.cc#L8)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Mac builder failure appears to be unrelated flake.  Merging.

---

Reviewed 1 of 7 files at r4, 2 of 2 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2172#-:-KGXJED_yngEXxe8fcKh:113868149)*
<!-- Sent from Reviewable.io -->
